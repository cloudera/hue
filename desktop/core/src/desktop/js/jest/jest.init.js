// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Ensure singletons
import 'core-js/stable';
import 'regenerator-runtime/runtime';

import 'apps/editor/execution/sessionManager';
import './jquery.setup';
import './sqlTestUtils';

import 'ext/bootstrap.2.3.2.min';
import axios from 'axios';
import $ from 'jquery';
import * as ko from 'knockout';
import komapping from 'knockout.mapping';

ko.mapping = komapping;

class Tooltip {}

class AceRange {}

class Autocomplete {}

const globalVars = {
  ko: ko,
  AUTOCOMPLETE_TIMEOUT: 1,
  CACHEABLE_TTL: 1,
  HAS_LINK_SHARING: true,
  HAS_SQL_ANALYZER: false,
  HUE_I18n: {},
  HUE_BASE_URL: '',
  HUE_CHARTS: {
    TYPES: 'bar'
  },
  LOGGED_USERNAME: 'foo',
  LOGGED_USER_ID: 'bar',
  PREVENT_AUTOFILL_INPUT_ATTRS:
    'autocorrect="off" autocomplete="do-not-autocomplete" autocapitalize="off" spellcheck="false"',
  STATIC_URLS: {
    'impala/art/icon_impala_48.png': 'impala/art/icon_impala_48.png',
    'beeswax/art/icon_beeswax_48.png': 'beeswax/art/icon_beeswax_48.png'
  },
  SQL_COLUMNS_KNOWN_FACET_VALUES: {
    type: { array: -1, boolean: -1 }
  },
  ace: {
    edit: () => ({
      setOptions: () => {},
      getSession: () => ({
        setMode: () => {},
        doc: {
          createAnchor: () => ({
            on: () => {},
            getPosition: () => ({
              row: 0
            }),
            setPosition: () => {},
            detach: () => {}
          })
        },
        getTextRange: () => {},
        addGutterDecoration: () => {},
        removeGutterDecoration: () => {}
      }),
      setTheme: () => {},
      getValue: () => '',
      getSelectionRange: () => ({ start: { row: 0, column: 0 }, end: { row: 0, column: 0 } }),
      on: () => {},
      off: () => {},
      commands: {
        on: () => {},
        off: () => {}
      },
      container: {
        addEventListener: () => {},
        removeEventListener: () => {}
      }
    }),
    require: () => ({
      Tooltip: Tooltip,
      Range: AceRange,
      Autocomplete: Autocomplete
    })
  }
};

Object.keys(globalVars).forEach(key => {
  global[key] = globalVars[key];
  global.window[key] = globalVars[key];
});

$.ajaxSetup({
  beforeSend: function () {
    console.warn('actual jQuery ajax called');
    console.trace();
  }
});

axios.interceptors.request.use(config => {
  console.warn('Actual axios ajax request made to url: ' + config.url);
  console.trace();
  return config;
});

process.on('unhandledRejection', err => {
  fail(err);
});

jest.mock('../utils/i18nReact');
jest.mock('../utils/hueAnalytics');
