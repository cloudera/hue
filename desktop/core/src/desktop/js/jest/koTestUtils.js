// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as ko from 'knockout';

export const koSetup = () => {
  let originalLoadTemplate;
  let wrapper;

  beforeEach(() => {
    originalLoadTemplate = ko.components.defaultLoader.loadTemplate;
    ko.components.defaultLoader.loadTemplate = (name, templateConfig, callback) => {
      const div = window.document.createElement('div');
      div.innerHTML = templateConfig;
      callback(div.children);
    };

    wrapper = window.document.createElement('div');
    wrapper.classList.add('component-wrapper');
    window.document.querySelector('body').appendChild(wrapper);
  });

  afterEach(() => {
    ko.components.defaultLoader.loadTemplate = originalLoadTemplate;
    wrapper.parentNode.removeChild(wrapper);
    ko.cleanNode(wrapper);
    expect(window.document.querySelectorAll('[data-test]').length).toEqual(0);
  });

  return {
    renderComponent: async (name, data, instantTimeout) =>
      new Promise((resolve, reject) => {
        const element = window.document.createElement('div');
        element.setAttribute(
          'data-bind',
          `descendantsComplete: descendantsComplete, component: { name: "${name}", params: params }`
        );
        wrapper.appendChild(element);

        try {
          ko.components.get(name, comp => {
            if (comp.createViewModel) {
              const origSetTimeout = window.setTimeout;
              if (instantTimeout) {
                window.setTimeout = cb => cb();
                comp.createViewModel(data, { element: element });
                window.setTimeout = origSetTimeout;
              } else {
                comp.createViewModel(data, { element: element });
              }

              ko.applyBindings(
                {
                  descendantsComplete: () => resolve(wrapper),
                  params: data
                },
                wrapper
              );
            } else {
              reject('no createViewModel function found on component ' + name);
            }
          });
        } catch (e) {
          reject(e);
        }
      }),

    renderKo: async (html, viewModel) =>
      new Promise(resolve => {
        const setupContainer = window.document.createElement('div');
        setupContainer.setAttribute(
          'data-bind',
          '{ descendantsComplete: descendantsComplete, with: model }'
        );

        const container = window.document.createElement('div');
        container.innerHTML = html;
        setupContainer.appendChild(container);

        const data = {
          model: viewModel,
          descendantsComplete: () => resolve(container)
        };

        ko.applyBindings(data, setupContainer);
      }),

    waitForKoUpdate: async () =>
      new Promise(resolve => {
        window.setTimeout(resolve, 0);
      })
  };
};

export const waitForObservableChange = async observable =>
  new Promise(resolve => {
    const sub = observable.subscribe(() => {
      sub.dispose();
      resolve();
    });
  });

export const simulateInput = (element, text) => {
  element.value = text;
  element.dispatchEvent(new Event('change'));
};
