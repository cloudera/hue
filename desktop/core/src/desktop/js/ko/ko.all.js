// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as ko from 'knockout';
import 'ko/ko.init';
import komapping from 'knockout.mapping';
import 'knockout-switch-case';
import 'knockout-sortable';
import 'knockout.validation';

import 'ext/ko.editable.custom';
import 'ext/ko.selectize.custom';

import 'ko/bindings/charts/mapchart/ko.mapChart';
import 'ko/bindings/charts/ko.barChart';
import 'ko/bindings/charts/ko.leafletMapChart';
import 'ko/bindings/charts/ko.lineChart';
import 'ko/bindings/charts/ko.partitionChart';
import 'ko/bindings/charts/ko.pieChart';
import 'ko/bindings/charts/ko.scatterChart';
import 'ko/bindings/charts/ko.timelineChart';

import 'ko/bindings/ko.vue.js';

import 'ko/bindings/ace/ko.aceEditor';
import 'ko/bindings/ace/ko.aceResizer';
import 'ko/bindings/ko.appAwareTemplateContextMenu';
import 'ko/bindings/ko.assistFileDraggable';
import 'ko/bindings/ko.assistFileDroppable';
import 'ko/bindings/ko.assistVerticalResizer';
import 'ko/bindings/ko.attachViewModelToElementData';
import 'ko/bindings/ko.augmentHtml';
import 'ko/bindings/ko.autocomplete';
import 'ko/bindings/ko.autogrowInput';
import 'ko/bindings/ko.blurHide';
import 'ko/bindings/ko.bytesize';
import 'ko/bindings/ko.chosen';
import 'ko/bindings/ko.clearable';
import 'ko/bindings/ko.clickForAceFocus';
import 'ko/bindings/ko.clickToCopy';
import 'ko/bindings/ko.clipboard';
import 'ko/bindings/ko.codemirror';
import 'ko/bindings/ko.contextMenu';
import 'ko/bindings/ko.contextSubMenu';
import 'ko/bindings/ko.datalist';
import 'ko/bindings/ko.datepicker';
import 'ko/bindings/ko.dateRangePicker';
import 'ko/bindings/ko.dblClick';
import 'ko/bindings/ko.delayedOverflow';
import 'ko/bindings/ko.dockable';
import 'ko/bindings/ko.documentChooser';
import 'ko/bindings/ko.documentContextPopover';
import 'ko/bindings/ko.draggableText';
import 'ko/bindings/ko.dropdown';
import 'ko/bindings/ko.dropzone';
import 'ko/bindings/ko.duration';
import 'ko/bindings/ko.ellipsis';
import 'ko/bindings/ko.fadeVisible';
import 'ko/bindings/ko.fetchMore';
import 'ko/bindings/ko.fileChooser';
import 'ko/bindings/ko.foreachVisible';
import 'ko/bindings/ko.fresherEditor';
import 'ko/bindings/ko.hdfsAutocomplete';
import 'ko/bindings/ko.hdfsTree';
import 'ko/bindings/ko.highlight';
import 'ko/bindings/ko.hiveChooser';
import 'ko/bindings/ko.html';
import 'ko/bindings/ko.hueach';
import 'ko/bindings/ko.hueCheckAll';
import 'ko/bindings/ko.hueCheckbox';
import 'ko/bindings/ko.hueChecked';
import 'ko/bindings/ko.hueLink';
import 'ko/bindings/ko.hueSpinner';
import 'ko/bindings/ko.impalaDagre';
import 'ko/bindings/ko.jHueRowSelector';
import 'ko/bindings/ko.logResizer';
import 'ko/bindings/ko.logScroller';
import 'ko/bindings/ko.medium';
import 'ko/bindings/ko.moment';
import 'ko/bindings/ko.momentFromNow';
import 'ko/bindings/ko.multiCheck';
import 'ko/bindings/ko.multiCheckForeachVisible';
import 'ko/bindings/ko.multiClick';
import 'ko/bindings/ko.multiLineEllipsis';
import 'ko/bindings/ko.numberFormat';
import 'ko/bindings/ko.numericTextInput';
import 'ko/bindings/ko.onClickOutside';
import 'ko/bindings/ko.oneClickSelect';
import 'ko/bindings/ko.parseArguments';
import 'ko/bindings/ko.publish';
import 'ko/bindings/ko.reactWrapper';
import 'ko/bindings/ko.readOnlyAce';
import 'ko/bindings/ko.resizable';
import 'ko/bindings/ko.select2';
import 'ko/bindings/ko.simplesize';
import 'ko/bindings/ko.slider';
import 'ko/bindings/ko.slideVisible';
import 'ko/bindings/ko.solrChooser';
import 'ko/bindings/ko.spinEdit';
import 'ko/bindings/ko.splitDraggable';
import 'ko/bindings/ko.splitFlexDraggable';
import 'ko/bindings/ko.sqlContextPopover';
import 'ko/bindings/ko.stickVertical';
import 'ko/bindings/ko.storageContextPopover';
import 'ko/bindings/ko.stretchDown';
import 'ko/bindings/ko.tagEditor';
import 'ko/bindings/ko.tagsNotAllowed';
import 'ko/bindings/ko.templateContextMenu';
import 'ko/bindings/ko.templatePopover';
import 'ko/bindings/ko.textSqueezer';
import 'ko/bindings/ko.toggle';
import 'ko/bindings/ko.toggleOverflow';
import 'ko/bindings/ko.tooltip';
import 'ko/bindings/ko.truncatedText';
import 'ko/bindings/ko.typeahead';
import 'ko/bindings/ko.verticalSlide';
import 'ko/bindings/ko.visibleOnHover';

import 'ko/components/appSwitcher/ko.appSwitcher';
import 'ko/components/assist/ko.assistDashboardPanel';
import 'ko/components/assist/ko.assistDbPanel';
import 'ko/components/assist/ko.assistDocumentsPanel';
import 'ko/components/assist/ko.assistEditorContextPanel';
import 'ko/components/assist/ko.assistFunctionsPanel';
import 'ko/components/assist/ko.assistGitPanel';
import 'ko/components/assist/ko.assistHBasePanel';
import 'ko/components/assist/ko.assistLangRefPanel';
import 'ko/components/assist/ko.assistPanel';
import 'ko/components/assist/ko.assistSchedulePanel';
import 'ko/components/assist/ko.assistStoragePanel';
import 'ko/components/assist/ko.rightAssistPanel';
import 'ko/components/contextPopover/ko.contextPopover';
import 'ko/components/contextPopover/ko.quickQueryContext';
import 'ko/components/contextPopover/ko.popoverOnHover';
import 'ko/components/doc/ko.shareDocModal';
import 'ko/components/simpleAceEditor/ko.simpleAceEditor';

import 'ko/components/ko.catalogEntriesList';
import 'ko/components/ko.contextSelector';
import 'ko/components/ko.createDirectoryModal';
import 'ko/components/ko.deleteDocModal';
import 'ko/components/ko.dropDown';
import 'ko/components/ko.editorDroppableMenu';
import 'ko/components/ko.executionAnalysis';
import 'ko/components/ko.favoriteApp';
import 'ko/components/ko.fieldSamples';
import 'ko/components/ko.globalSearch';
import 'ko/components/ko.historyPanel';
import 'ko/components/ko.importDocumentsModal';
import 'ko/components/ko.inlineAutocomplete';
import 'ko/components/ko.jobBrowserLinks';
import 'ko/components/ko.navProperties';
import 'ko/components/ko.navTags';
import 'ko/components/ko.pollingCatalogEntriesList';
import 'ko/components/ko.quickQueryAction';
import 'ko/components/ko.sentryPrivileges';
import 'ko/components/ko.sessionAuthModal';
import 'ko/components/ko.sessionPanel';
import 'ko/components/ko.shareGistModal';
import 'ko/components/ko.sqlColumnsTable';

// TODO: Move to about app when it has it's own webpack entry
import 'apps/about/components/ko.hueConfigTree';
import 'apps/about/components/ko.connectorsConfig';

import 'ko/extenders/ko.maxLength';
import 'ko/extenders/ko.numeric';
import 'ko/extenders/ko.toJson';

import 'ko/observables/ko.observableArrayDefault';
import 'ko/observables/ko.observableDefault';

window.ko = ko;
window.ko.mapping = komapping;

export default ko;
