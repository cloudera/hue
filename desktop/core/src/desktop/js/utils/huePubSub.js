// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Based on original pub/sub implementation from http://davidwalsh.name/pubsub-javascript

const topics = {};
const hOP = topics.hasOwnProperty;

const huePubSub = {
  subscribe: function (topic, listener, app) {
    if (!hOP.call(topics, topic)) {
      topics[topic] = [];
    }

    const index =
      topics[topic].push({
        listener: listener,
        app: app,
        status: 'running'
      }) - 1;

    return {
      remove: function () {
        delete topics[topic][index];
      }
    };
  },
  removeAll: function (topic) {
    topics[topic] = [];
  },
  subscribeOnce: function (topic, listener, app) {
    const ephemeral = this.subscribe(
      topic,
      function () {
        listener.apply(listener, arguments);
        ephemeral.remove();
      },
      app
    );
  },
  publish: function (topic, info) {
    if (!hOP.call(topics, topic)) {
      return;
    }

    topics[topic].forEach(item => {
      if (item.status === 'running') {
        item.listener(info);
      }
    });
  },
  getTopics: function () {
    return topics;
  },
  pauseAppSubscribers: function (app) {
    if (app) {
      Object.keys(topics).forEach(topicName => {
        topics[topicName].forEach(topic => {
          if (
            typeof topic.app !== 'undefined' &&
            topic.app !== null &&
            (topic.app === app || topic.app.split('-')[0] === app)
          ) {
            topic.status = 'paused';
          }
        });
      });
    }
  },
  resumeAppSubscribers: function (app) {
    if (app) {
      Object.keys(topics).forEach(topicName => {
        topics[topicName].forEach(topic => {
          if (
            typeof topic.app !== 'undefined' &&
            topic.app !== null &&
            (topic.app === app || topic.app.split('-')[0] === app)
          ) {
            topic.status = 'running';
          }
        });
      });
    }
  },
  clearAppSubscribers: function (app) {
    if (app) {
      Object.keys(topics).forEach(topicName => {
        topics[topicName] = topics[topicName].filter(obj => {
          return obj.app !== app;
        });
      });
    }
  }
};

export default huePubSub;
