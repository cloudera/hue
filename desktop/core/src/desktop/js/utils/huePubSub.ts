// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Based on original pub/sub implementation from http://davidwalsh.name/pubsub-javascript

/* eslint-disable @typescript-eslint/no-explicit-any */

interface Listener<T> {
  (data: T): void;
}

enum PubSubState {
  RUNNING,
  PAUSED
}

const topics: {
  [topic: string]: {
    listener: Listener<any>;
    app?: string;
    state: PubSubState;
  }[];
} = {};

const hOP = topics.hasOwnProperty;

export interface HueSubscription {
  remove: () => void;
}

const subscribe = <T = any>(
  topic: string,
  listener: Listener<T>,
  app?: string
): HueSubscription => {
  if (!hOP.call(topics, topic)) {
    topics[topic] = [];
  }

  const index =
    topics[topic].push({
      listener: listener,
      app: app,
      state: PubSubState.RUNNING
    }) - 1;

  return {
    remove: () => {
      delete topics[topic][index];
    }
  };
};

const removeAll = (topic: string): void => {
  topics[topic] = [];
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
const publish = <T = any>(topic: string, info?: T): void => {
  if (!hOP.call(topics, topic)) {
    return;
  }

  topics[topic].forEach(item => {
    if (item.state === PubSubState.RUNNING) {
      item.listener(info);
    }
  });
};

const subscribeOnce = <T = any>(
  topic: string,
  listener: Listener<T>,
  app: string
): HueSubscription => {
  const ephemeral = subscribe(
    topic,
    arg => {
      listener(arg);
      ephemeral.remove();
    },
    app
  );
  return ephemeral;
};

const getTopics = (): typeof topics => topics;

const setStatusForApp = (state: PubSubState, app?: string): void => {
  if (app) {
    Object.keys(topics).forEach(topicName => {
      topics[topicName].forEach(topic => {
        if (topic.app && (topic.app === app || topic.app.split('-')[0] === app)) {
          topic.state = state;
        }
      });
    });
  }
};

const pauseAppSubscribers = (app?: string): void => {
  setStatusForApp(PubSubState.PAUSED, app);
};

const resumeAppSubscribers = (app?: string): void => {
  setStatusForApp(PubSubState.RUNNING, app);
};

const clearAppSubscribers = (app?: string): void => {
  if (app) {
    Object.keys(topics).forEach(topicName => {
      topics[topicName] = topics[topicName].filter(obj => {
        return obj.app !== app;
      });
    });
  }
};

export default {
  clearAppSubscribers,
  getTopics,
  pauseAppSubscribers,
  publish,
  removeAll,
  resumeAppSubscribers,
  subscribe,
  subscribeOnce
};
