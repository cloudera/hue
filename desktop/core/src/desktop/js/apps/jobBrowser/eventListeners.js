// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import $ from 'jquery';
import huePubSub from '../../utils/huePubSub';
import { GLOBAL_ERROR_TOPIC, GLOBAL_INFO_TOPIC } from 'reactComponents/AlertComponent/events';

$(document).off('shown', '.jb-logs-link');
$(document).on('shown', '.jb-logs-link', e => {
  const dest = $(e.target).attr('href');
  if (dest.indexOf('logs') > -1 && $(dest).find('pre:visible').length > 0) {
    $(dest)
      .find('pre')
      .css('overflow-y', 'auto')
      .height(
        Math.max(
          200,
          $(window).height() -
            $(dest).find('pre').offset().top -
            $('.page-content').scrollTop() -
            75
        )
      );
  }
});

$(document).off('showSubmitPopup');
$(document).on('showSubmitPopup', (event, data) => {
  const syncWorkflowModal = $('#syncWorkflowModal');
  syncWorkflowModal.empty();
  syncWorkflowModal.html(data);
  syncWorkflowModal.modal('show');
  syncWorkflowModal.on('hidden', () => {
    huePubSub.publish('hide.datepicker');
  });

  syncWorkflowModal.find('.submit-form').on('submit', function (e) {
    e.preventDefault();
    $.ajax({
      type: 'POST',
      cache: false,
      url: $(this).attr('action'),
      data: $(this).serialize(),
      success: function (data) {
        $('#syncWorkflowModal').modal('hide');
        if (data && data.status === 0) {
          huePubSub.publish(GLOBAL_INFO_TOPIC, data);
        } else {
          huePubSub.publish(GLOBAL_ERROR_TOPIC, data);
        }
      },
      error: function (data) {
        $('#syncWorkflowModal').modal('hide');
        huePubSub.publish(GLOBAL_ERROR_TOPIC, data);
      }
    });
  });
});
