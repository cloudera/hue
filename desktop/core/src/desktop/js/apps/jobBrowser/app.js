// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import $ from 'jquery';
import * as ko from 'knockout';

import huePubSub from 'utils/huePubSub';

import './components/hiveQueryPlan/webcomp';
import './components/queriesList/webcomp';
import './components/impalaQueries/webcomp';
import './eventListeners';
import JobBrowserViewModel from './knockout/JobBrowserViewModel';
import Job from './knockout/Job';
import { GLOBAL_ERROR_TOPIC, GLOBAL_INFO_TOPIC } from 'reactComponents/AlertComponent/events';
import I18n from 'utils/i18n';

huePubSub.subscribe('app.dom.loaded', app => {
  if (app !== 'jobbrowser') {
    return;
  }

  $(document).ready(() => {
    const jobBrowserViewModel = new JobBrowserViewModel(false);
    const openJob = id => {
      if (jobBrowserViewModel.job() == null) {
        jobBrowserViewModel.job(new Job(jobBrowserViewModel, {}));
      }
      jobBrowserViewModel.job().id(id);
      jobBrowserViewModel.job().fetchJob();
    };

    ko.applyBindings(jobBrowserViewModel, $('#jobbrowserComponents')[0]);

    const loadHash = () => {
      if (window.location.pathname.indexOf('jobbrowser') > -1) {
        jobBrowserViewModel.load();
      }
    };

    window.onhashchange = () => loadHash();

    huePubSub.subscribe(
      'oozie.action.logs.click',
      widget => {
        $.get(
          widget.logsURL(),
          {
            format: 'link'
          },
          data => {
            const id = data.job || data.attemptid;
            if (id) {
              openJob(id);
            } else {
              huePubSub.publish(GLOBAL_ERROR_TOPIC, { message: I18n('No log available') });
            }
          }
        );
      },
      'jobbrowser'
    );

    huePubSub.subscribe(
      'oozie.action.click',
      widget => {
        openJob(widget.externalId());
      },
      'jobbrowser'
    );

    huePubSub.subscribe(
      'browser.job.open.link',
      id => {
        openJob(id);
      },
      'jobbrowser'
    );

    huePubSub.subscribe(
      'app.gained.focus',
      app => {
        if (app === 'jobbrowser') {
          huePubSub.publish('graph.draw.arrows');
          loadHash();
        }
      },
      'jobbrowser'
    );

    const configUpdated = clusterConfig => {
      jobBrowserViewModel.appConfig(clusterConfig && clusterConfig['app_config']);
      jobBrowserViewModel.clusterType(clusterConfig && clusterConfig['cluster_type']);
      loadHash();
    };

    huePubSub.subscribe('cluster.config.set.config', configUpdated);
    huePubSub.publish('cluster.config.get.config', configUpdated);

    huePubSub.subscribe(
      'submit.rerun.popup.return',
      data => {
        huePubSub.publish(GLOBAL_INFO_TOPIC, { message: I18n('Rerun submitted.') });
        $('#rerun-modal').modal('hide');

        jobBrowserViewModel.job().apiStatus('RUNNING');
        jobBrowserViewModel.monitorJob(jobBrowserViewModel.job());
      },
      'jobbrowser'
    );
  });
});
