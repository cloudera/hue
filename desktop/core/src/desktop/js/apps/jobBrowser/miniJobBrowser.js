// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import $ from 'jquery';
import * as ko from 'knockout';

import { GLOBAL_INFO_TOPIC } from 'reactComponents/AlertComponent/events';
import huePubSub from 'utils/huePubSub';
import I18n from 'utils/i18n';

import './components/hiveQueryPlan/webcomp';
import './components/queriesList/webcomp';
import './components/impalaQueries/webcomp';
import './eventListeners';
import JobBrowserViewModel from './knockout/JobBrowserViewModel';
import Job from './knockout/Job';

export const initializeMiniJobBrowser = () => {
  const jobBrowserViewModel = new JobBrowserViewModel(true);
  const openJob = id => {
    if (jobBrowserViewModel.job() == null) {
      jobBrowserViewModel.job(new Job(jobBrowserViewModel, {}));
    }
    jobBrowserViewModel.job().id(id);
    jobBrowserViewModel.job().fetchJob();
  };

  ko.applyBindings(jobBrowserViewModel, $('#jobbrowserMiniComponents')[0]);

  const configUpdated = clusterConfig => {
    jobBrowserViewModel.appConfig(clusterConfig && clusterConfig['app_config']);
    jobBrowserViewModel.clusterType(clusterConfig && clusterConfig['cluster_type']);
  };

  huePubSub.subscribe('cluster.config.set.config', configUpdated);
  huePubSub.publish('cluster.config.get.config', configUpdated);

  huePubSub.subscribe('submit.rerun.popup.return-mini', () => {
    huePubSub.publish(GLOBAL_INFO_TOPIC, { message: I18n('Rerun submitted.') });
    $('#rerun-modal-mini').modal('hide');

    jobBrowserViewModel.job().apiStatus('RUNNING');
    jobBrowserViewModel.monitorJob(jobBrowserViewModel.job());
  });

  huePubSub.subscribe('mini.jb.navigate', options => {
    if (options.compute) {
      jobBrowserViewModel.compute(options.compute);
    }
    $('#jobsPanel .nav-pills li').removeClass('active');
    const interfaceName = jobBrowserViewModel.isValidInterface(options.section);
    $('#jobsPanel .nav-pills li[data-interface="' + interfaceName + '"]').addClass('active');
    jobBrowserViewModel.selectInterface(interfaceName);
  });

  huePubSub.subscribe('mini.jb.open.job', openJob);

  huePubSub.subscribe('mini.jb.expand', () => {
    if (jobBrowserViewModel.job()) {
      huePubSub.publish('open.link', '/jobbrowser/#!id=' + jobBrowserViewModel.job().id());
    } else {
      huePubSub.publish('open.link', '/jobbrowser/#!' + jobBrowserViewModel.interface());
    }
  });
};
