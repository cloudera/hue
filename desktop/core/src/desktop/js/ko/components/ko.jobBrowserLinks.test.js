// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import $ from 'jquery';
import { koSetup } from 'jest/koTestUtils';
import { NAME } from './ko.jobBrowserLinks';
import huePubSub from 'utils/huePubSub';

import { initializeMiniJobBrowser } from 'apps/jobBrowser/miniJobBrowser';
jest.mock('apps/jobBrowser/miniJobBrowser');

describe('ko.jobBrowserLinks.js', () => {
  const setup = koSetup();
  let fakeMiniPanel;

  beforeEach(() => {
    jest.spyOn($, 'post').mockImplementation(url => {
      if (url === '/jobbrowser/jobs/') {
        return $.Deferred().resolve({ jobs: [] }).promise();
      }
    });

    // This element gets created during runtime via an ajax call that gets the raw
    // html contents including scripts and styles.
    fakeMiniPanel = document.createElement('div');
    fakeMiniPanel.setAttribute('id', 'jobsPanel');
    document.body.appendChild(fakeMiniPanel);
  });

  afterEach(() => {
    jest.clearAllMocks();
    fakeMiniPanel.remove();
  });

  it('should render component', async () => {
    const element = await setup.renderComponent(NAME, {});

    expect(element.innerHTML).toMatchSnapshot();
  });

  it('should initialize the mini job browser on first open', async () => {
    window.HAS_JOB_BROWSER = true;
    window.getLastKnownConfig = () => ({});
    jest.spyOn($, 'post').mockImplementation(url => {
      if (url === '/jobbrowser/jobs/') {
        return Promise.resolve({ jobs: [] }).promise();
      }
    });
    await setup.renderComponent(NAME, {});
    expect(initializeMiniJobBrowser.mock.calls).toHaveLength(0);

    // First open
    huePubSub.publish('show.jobs.panel');
    expect(initializeMiniJobBrowser.mock.calls).toHaveLength(1);

    huePubSub.publish('hide.jobs.panel');

    // Second open
    huePubSub.publish('show.jobs.panel');
    expect(initializeMiniJobBrowser.mock.calls).toHaveLength(1);
  });
});
