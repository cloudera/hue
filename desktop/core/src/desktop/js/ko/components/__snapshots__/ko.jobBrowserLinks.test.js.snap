// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ko.jobBrowserLinks.js should render component 1`] = `
"<div data-bind=\\"descendantsComplete: descendantsComplete, component: { name: &quot;hue-job-browser-links&quot;, params: params }\\"><div class=\\"btn-group pull-right\\">
    <!-- ko if: window.IS_K8S_ONLY --><!-- /ko -->
    <!-- ko if: window.HAS_JOB_BROWSER --><!-- /ko -->
  </div></div>"
`;
